subsys/net/lib/tftp/tftp_client.c

```diff
@@ -247,7 +308,87 @@ int tftp_get(struct sockaddr *server_addr, struct tftpc *client,
 		ret = TFTPC_REMOTE_ERROR;
 	}
 
-req_abort:
+get_abort:
+	k_mutex_unlock(&tftpc_lock);
+	close(sock);
+	return ret;
+}
+
+int tftp_put(struct sockaddr *server_addr, struct tftpc *client,
+	     const char *remote_file, const char *mode)
+{
+	int sock;
+	uint32_t tftpc_block_no = 1;
+	uint32_t tftpc_index = 0;
+	int ret;
+
+	if (client->user_buf == NULL || client->user_buf_size == 0) {
+		return -EFAULT;
+	}
+
+	sock = socket(server_addr->sa_family, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		LOG_ERR("Failed to create UDP socket (IPv4): %d", errno);
+		return -errno;
+	}
+
+	/* Obtain Global Lock before accessing critical resources. */
+	k_mutex_lock(&tftpc_lock, K_FOREVER);
+
+	/* Send out the WRITE request to the TFTP Server. */
+	ret = send_request(sock, server_addr, WRITE_REQUEST, remote_file, mode);
+
+	/* Check connection initiation result */
+	if (ret >= TFTP_HEADER_SIZE) {
+		uint16_t opcode = sys_get_be16(tftpc_buffer);
+		uint16_t block_no = sys_get_be16(tftpc_buffer + 2);
+
+		LOG_DBG("Receive: opcode %u, block no %u, size %d", opcode, block_no, ret);
+
+		if (opcode == ERROR_OPCODE) {
+			LOG_ERR("Server responded with service reject.");
+			return TFTPC_REMOTE_ERROR;
+		} else if (opcode != ACK_OPCODE || block_no != 0) {
+			LOG_ERR("Server responded with invalid opcode or block number.");
+			ret = TFTPC_REMOTE_ERROR;
+			goto put_abort;
+		}
+	} else {
+		ret = TFTPC_REMOTE_ERROR;
+		goto put_abort;
+	}
+
+
+	do {
+		uint8_t send_buffer[TFTPC_MAX_BUF_SIZE];
```

Comment:
This looks very dangerous to do in a library code as the buf size can be > 512 bytes, which means that you might be overwrite the stack. Please re-architecture your code to avoid this. One option is to let the caller provide a temp buffer for this function and use that.

Reply:
Based on the protocol, send and receive does not happen simultaneously.
Reuse the existing `tftpc_buffer` to send out data in 512-byte chunks.

Reply:
That is indeed better but this library is not really a proper library because the buffer is shared between possible API callers. Perhaps this is not a real issue in the applications (that the API is not re-entrant), but perhaps this could be addressed at some point (fortunately the API is marked as experimental so should be easy to change if needed).

Reply:
The mutex could help a bit, but the real problem is that one same thread could lock a mutex multiple times.